keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,158 @@
+
+## Motivation
+
+### The Problems
+
+From our testing, the problems seem to be because of a large number of small object allocations.
+
+#### On list, update, and apply: 
+
+Deserializing managedFields as a deeply nested field in the apiserver allocates many small objects. Objects are deserialized when fetched from etcd.
+The size of managedFields is similar to the size of the object. It increases etcd size, and cache size of the apiserver and clients.
+
+#### On update and apply: 
+
+Converting managedFields from the api representation into the internal representation causes many small object allocations.
+Some structured-merge-diff operations (like compare) also create many small objects that are not needed. Most “field set” operations are not done in place.
+
+### Experiments
+
+The following experiments convinced us that the problem is mostly the allocations of small objects. Latency in etcd because of read/write with bigger objects, and/or data transfer was not a significant issue.
+
+For all these scenarios, we’ve created a deployment with 5,000 pods. Pods that have the managedFields (when server-side apply is enabled) are obviously larger. Here are some numbers for the comparison, with the objects in protobuf format:
+
+- Server-side apply disabled:
+  - A pod is about 8,137 bytes
+  - 5,000 pods is 40,506,271 bytes
+- Server-side apply enabled:
+  - A pod is about 13,164 bytes (~62% larger)
+  - 5,000 pods is 67,182,690 bytes
+
+All the following scenarios have been tested by LISTing all the pods, then running 250 patches per second while still LISTing the pods at the same time.
+
+1. Server-side apply disabled, 
+1. Server-side apply enabled, 
+1. Server-side apply disabled, but with an extra annotation. The annotation is about 5,000 bytes to compensate for the difference in object size with server-side apply.
+1. Server-side apply disabled, with 320 small annotations. Those annotations total about 5,000 bytes to simulate the small object allocation.
+1. Server-side apply enabled, with metav1.Fields as a string instead of a nested map (So it doesn't have to be deserialized on LIST). The content of the string is serialized JSON of the same nested map.
+1. Server-side apply enabled, with metav1.Fields as a string instead of a nested map. The content of the string is a list of full field paths, instead of a trie.
+
+The average latency are as follows:
+
+| Scenario | Object Size | LIST | PATCH | LIST while PATCH |
+|---|---|---|---|---|
+| 1. | 8.1 KB | 360.4 ms | 8.0 ms | 361.0 ms |
+| 2. | 13.2 KB | 1294.5 ms | 38.5 ms | 2529.6 ms |
+| 3. | 13.2 KB | 480.8 ms | 11.3 ms | 500.2 ms |
+| 4. | 13.2 KB | 1260.1 ms | 9.7 ms | 1503.2 ms |
+| 5. | 13.7 KB | 544.0 ms | 26.6 ms | 735.1 ms |
+| 6. | 27.2 KB | 1791.3 ms | 39.7 ms | 3803.4 ms |
```

Comment:
We did update this table with new numbers, which I think are pretty encouraging.

Reply:
So apparently that is much worse - is it because of a significant nesting in pod?

So going back to point 5 (simple serialization as list), is there any way to reduce the size of this map further?
I'm assuming that in 99% of cases, majority of spec will be written by a single actor, whole status written by kubelet, nodeName by scheduler, and that's mostly it, right?
Having additional 5kB of data to pass that information seems like quite huge.

The results look promissing enough that I'm not that afraid about some corner cases where all fields are written by different parties, but in majority of cases I would expect it to be like I described above.

@deads2k - for thoughts too

Reply:
Looking at the result of 5 compared with 3, I think I'd look at trying to close the 50% bump in "list while patch" or the 2.5x bump in the patch performance before trying to shrink the size.  Can the difference be reproduced in a benchmark where you have good shot at seeing where it comes from?

Reply:
@jennybuckley can we update this with the latest numbers?